torch/_functorch/autograd_function.py

```diff
@@ -387,29 +391,22 @@ def custom_function_call_functionalize(
 
 
 def vmapify_autograd_function(autograd_function, in_dims, batch_size, randomness):
-    # The following values are saved from the forward() and setup_context()
-    # and used in backward().
-    # Why do we save the values out here instead of on the ctx object?
-    # - out_dims: There's no way to retrieve this from forward()
-    # - input_shapes, saved_tensors_bdims: I'm a bit scared of nesting
-    #   vmap(vmap( but not completely sure if it is a problem. If we
-    #   assigned those fields to the ctx object, the worry is that they
-    #   get overwritten.
-    init_val = "not populated"
-    out_dims = init_val
-    input_shapes: Any = init_val
-    saved_tensors_bdims: Any = init_val
-
     def forward(*operands):
-        nonlocal out_dims
         outputs, out_dims = restore_vmap(
             autograd_function.forward, in_dims, batch_size, randomness
         )(*operands)
-        return outputs
+        if isinstance(outputs, torch.Tensor):
+            return outputs, out_dims
+        else:
+            return *outputs, out_dims
 
     def setup_context(ctx, inputs, outputs):
-        input_shapes_ = None
-        saved_tensors_bdims_ = None
+        out_dims = outputs[-1]
+        if isinstance(out_dims, tuple):
+            outputs_ = outputs[:-1]
+        else:
+            outputs_ = outputs[0]
```

Comment:
Can you put this into a helper function? unpack_outputs(outputs) -> outputs, out_dims. 

This is used at least twice and I had to really think about it the first time. The reason why it checks if out_dims is a tuple is that if it is not a tuple, then that means the original output was a single Tensor